clients/hadoopfs/src/test/java/io/lakefs/LakeFSFileSystemTest.java

```diff
@@ -289,41 +288,215 @@ public void testAppend() throws IOException {
+        ObjectLocation srcDirObjLoc = fs.pathToObjectLocation(srcDir);
+        mockExistingDirPath(srcDirObjLoc, new ArrayList<ObjectLocation>(Arrays.asList(fileObjLoc)));
+
+        Path dst = new Path("lakefs://repo/main/non-existing-dir");
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+        mockNonExistingPath(dstObjLoc);
+
+        boolean renamed = fs.rename(srcDir, dst);
+        Assert.assertTrue(renamed);
+        Path expectedDstPath = new Path("lakefs://repo/main/non-existing-dir/existing.src");
+        Assert.assertTrue(dstPathLinkedToSrcPhysicalAddress(fileObjLoc, fs.pathToObjectLocation(expectedDstPath)));
+        verifyObjDeletion(fileObjLoc);
+    }
+
+    /**
+     * rename(srcDir(containing srcDir/a.txt), existing-dstdir) -> existing-dstdir/srcDir/a.txt
+     */
+    @Test
+    public void testRename_existingDirToExistingDirName() throws ApiException, IOException {
+        Path firstSrcFile = new Path("lakefs://repo/main/existing-dir1/a.src");
+        ObjectLocation firstObjLoc = fs.pathToObjectLocation(firstSrcFile);
+        Path secSrcFile = new Path("lakefs://repo/main/existing-dir1/b.src");
+        ObjectLocation secObjLoc = fs.pathToObjectLocation(secSrcFile);
+
+        Path srcDir = new Path("lakefs://repo/main/existing-dir1");
+        ObjectLocation srcDirObjLoc = fs.pathToObjectLocation(srcDir);
+        mockExistingDirPath(srcDirObjLoc, new ArrayList<ObjectLocation>(Arrays.asList(firstObjLoc, secObjLoc)));
+
+        Path fileInDstDir = new Path("lakefs://repo/main/existing-dir2/file.dst");
+        ObjectLocation dstFileObjLoc = fs.pathToObjectLocation(fileInDstDir);
+        Path dstDir = new Path("lakefs://repo/main/existing-dir2");
+        ObjectLocation dstDirObjLoc = fs.pathToObjectLocation(dstDir);
+        mockExistingDirPath(dstDirObjLoc, new ArrayList<ObjectLocation>(Arrays.asList(dstFileObjLoc)));
+
+        boolean renamed = fs.rename(srcDir, dstDir);
+        Assert.assertTrue(renamed);
+        Path firstExpectedDst = new Path("lakefs://repo/main/existing-dir2/existing-dir1/a.src");
+        Path secExpectedDst = new Path("lakefs://repo/main/existing-dir2/existing-dir1/b.src");
+        Assert.assertTrue(dstPathLinkedToSrcPhysicalAddress(firstObjLoc, fs.pathToObjectLocation(firstExpectedDst)));
+        Assert.assertTrue(dstPathLinkedToSrcPhysicalAddress(secObjLoc, fs.pathToObjectLocation(secExpectedDst)));
+        verifyObjDeletion(firstObjLoc);
+        verifyObjDeletion(secObjLoc);
+    }
+
+    @Test
+    public void testRename_srcAndDstOnDifferentBranch() throws IOException {
+        Path src = new Path("lakefs://repo/branch/existing.src");
+        ObjectLocation srcObjLoc = fs.pathToObjectLocation(src);
+
+        Path dst = new Path("lakefs://repo/another-branch/existing.dst");
+        ObjectLocation dstObjLoc = fs.pathToObjectLocation(dst);
+
+        boolean renamed = fs.rename(src, dst);
+        Assert.assertFalse(renamed);
```

Comment:
Here we might also want to verify that no API operations were performed on staging.